libnmstate/nm/device.py

```diff
@@ -49,6 +49,93 @@ def delete(dev):
         con_profile.delete()
 
 
+def reapply(dev, connection_profile):
```

Comment:
nit: Seems this method could be tested in a unit test.

Reply:
Why is the reapply function part of the device module? It seems it would be a easier to change the activation code in the connection module to try reapply and then fallback to normal activation. Then the caller does not need to decide whether or not to use reapply. This would also have the advantage of not adding so much non object-oriented code.

Reply:
Since you expect the feature in NM, I do not follow why Nmstate should not be better than NM in this case. I did no mean to suggest to try activation first and then reapply. My suggestion is to provide one interface to activate a connection for users of the nm submodule and make this one try a reapply first and fallback to normal activation.

Reply:
My comment is about not requesting users of the nm module to know that they need to use reapply but just make the activate method in the connection module do the right thing (trying to use reapply and activate if necessary).

Reply:
I feel we are in a loop here.
NM has two api/s, activate and reapply, it is a fact.
I decided to introduce the reapply one with a fallback addition.
You want me to change activate and not introduce reapply.
I explained that it does not matter how we call it and trying to change activate is a mess because I cannot depend there on the device module. I even explained why the reapply is in the device module and to the connection.

Further down, based on your remarks about the tests and special handling, I suggested we keep the two nm api/s as is and use a 3rd one with the fallback solving the test problem.

Please be explicit in what you suggest now base on the above info.

Reply:
The reason why you want to add the reapply code to nm.device is because NM uses it in a device. IMHO it does not matter what NM does if we want to do something better. If adding the code to the device module creates a problem because of a dependency loop to provide the better API, then IMHO the answer is to move the code elsewhere. It could be put as-is into the connection module or it could be put into its own module.

Also splitting the functions does not solve the test problem, because the test will allow to change the code to make all all two or three functions behave the same and it would not fail as long as there is not actual functional check that traffic is not disturbed. Splitting the functions could be use to keep the code in the matching nm modules and also provide a module with a better API than the low-level modules. Since we are time-constraint it still seems to me to be the better intermediate solution to add the code just to the connection module because creating a new module layer in a nice way would probably take some time. Otherwise it will just be shallow code.

You have your opinion, I have mine. I do not block the review on this but you did not persuade me that your solution is better.